crates/libcgroups/src/v1/freezer.rs

```diff
@@ -114,17 +122,43 @@ impl Freezer {
         Ok(())
     }
 
-    fn read_freezer_state(cgroup_root: &Path) -> Result<String, WrappedIoError> {
+    pub fn read_freezer_state(cgroup_root: &Path) -> FreezerState {
         let path = cgroup_root.join(CGROUP_FREEZER_STATE);
-        let mut content = String::new();
-        OpenOptions::new()
-            .create(false)
-            .read(true)
-            .open(path)
-            .wrap_open(cgroup_root)?
-            .read_to_string(&mut content)
-            .wrap_read(cgroup_root)?;
-        Ok(content)
+        loop {
+            let state = match read_to_string(&path) {
+                Ok(state) => state,
+                Err(_) => {
+                    return FreezerState::Undefined;
+                }
+            };
+            match state.trim() {
+                FREEZER_STATE_THAWED => return FreezerState::Thawed,
+                FREEZER_STATE_FROZEN => {
+                    // Find out whether the cgroup is frozen directly,
+                    // or indirectly via an ancestor.
+                    let self_state = match read_to_string(path.join(CGROUP_FREEZER_SELF_STATE)) {
+                        Ok(self_state) => self_state,
+                        Err(_) => {
+                            return FreezerState::Frozen;
+                        }
+                    };
+                    return match self_state.as_str() {
+                        "0\n" => FreezerState::Thawed,
+                        "1\n" => FreezerState::Frozen,
+                        _ => FreezerState::Undefined,
+                    };
+                }
+                FREEZER_STATE_FREEZING => {
+                    // Make sure we get a stable freezer state, so retry if the cgroup
+                    // is still undergoing freezing. This should be a temporary delay.
+                    thread::sleep(time::Duration::from_millis(1));
+                    continue;
+                }
```

Comment:
Technically this will exit after a couple of tries, but I feel we should still have a counter and return undefined/frozen after n-retries, just so we can be sure we'll never get stuck in infinite loop